src/main/rx/crsf.c

```diff
@@ -114,6 +116,107 @@ struct crsfPayloadRcChannelsPacked_s {
+* int8_t Uplink SNR ( db )
+* uint8_t Diversity active antenna ( enum ant. 1 = 0, ant. 2 )
+* uint8_t RF Mode ( enum 4fps = 0 , 50fps, 150hz)
+* uint8_t Uplink TX Power ( enum 0mW = 0, 10mW, 25 mW, 100 mW, 500 mW, 1000 mW, 2000mW )
+* uint8_t Downlink RSSI ( dBm * -1 )
+* uint8_t Downlink package success rate / Link quality ( % )
+* int8_t Downlink SNR ( db )
+* Uplink is the connection from the ground to the UAV and downlink the opposite direction.
+*/
+
+typedef struct crsfPayloadLinkstatistics_s {
+
+    uint8_t Uplink_RSSI_1;
+    uint8_t Uplink_RSSI_2;
+    uint8_t Uplink_Link_quality;
+    uint8_t Uplink_SNR;
+    uint8_t active_antenna;
+    uint8_t RF_Mode;
+    uint8_t Uplink_TX_Power;
+    uint8_t Downlink_RSSI;
+    uint8_t Downlink_Link_quality;
+    uint8_t Downlink_SNR;
+
+}crsfPayloadLinkstatistics_t;
+
+static crsfPayloadLinkstatistics_t crsfRXLinkstatistics;
+static uint32_t LastLinkStatisticsFrameUs;
+
+int converdbm( int x){
+    x = x * -1;
+    if (x < -92){
+        return 1;
+    }
+
+    if (x > -21){
+        return 100;
+    }
+    return  (int)(((-0.0154*x*x)-(0.3794*x)+98.182) + 0.5f);
+
+}
+
+void  handleCrsfLinkStatisticsFrame(uint8_t *frameStart,timeUs_t currentTimeUs)
+{
+    LastLinkStatisticsFrameUs = currentTimeUs;
+    memcpy(&crsfRXLinkstatistics, frameStart, CRSF_FRAME_LINK_STATISTICS_PAYLOAD_SIZE);
+
+    switch (debugMode)
+    {
+       case DEBUG_CRSF_LINK_STATISTICS_UPLINK:{
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 0, crsfRXLinkstatistics.Uplink_RSSI_1);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 1, crsfRXLinkstatistics.Uplink_RSSI_2);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 2, crsfRXLinkstatistics.Uplink_Link_quality);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 3, crsfRXLinkstatistics.RF_Mode);
+
+       break;
+       }
+       case DEBUG_CRSF_LINK_STATISTICS_PWR:{
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 0, crsfRXLinkstatistics.active_antenna);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 1, crsfRXLinkstatistics.Uplink_SNR);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 2, crsfRXLinkstatistics.Uplink_TX_Power);
+
+       break;
+       }
+       case DEBUG_CRSF_LINK_STATISTICS_DOWN:{
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 0, crsfRXLinkstatistics.Downlink_RSSI);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 1, crsfRXLinkstatistics.Downlink_Link_quality);
+           DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 2, crsfRXLinkstatistics.Downlink_SNR);
+
+       break;
+       }
+
+    }
+
+    int  rssidbm = (crsfRXLinkstatistics.active_antenna == 0 )? crsfRXLinkstatistics.Uplink_RSSI_1 :crsfRXLinkstatistics.Uplink_RSSI_2;
+
+    setRssiDirect(scaleRange(converdbm(rssidbm), 1,100, 0, RSSI_MAX_VALUE),RSSI_SOURCE_RX_PROTOCOL_CRFS);
+
+    setLinkQualityDirect( (crsfRXLinkstatistics.RF_Mode == 2) ? (100 + crsfRXLinkstatistics.Uplink_Link_quality) : crsfRXLinkstatistics.Uplink_Link_quality);
+
+}
```

Comment:
i get the idea but the code has issues 
lastTimeUs = currentTimeUs;  is called twice .. 
and 
static timeUs_t lastTimeUs = currentTimeUs; 
 error: initializer element is not constant

ill work on it

Reply:
It's just a suggestion so it could have issues.  Just trying to share the spirit of the concept is all.  you could declare it without initialization and the second cycle should pick up normally.

Reply:
Cleaned it up a bit.. is this better?  This assumes you're keeping `static timeUs_t lastLinkStatisticsFrameUs` outside of the handler

```
static void handleCrsfLinkStatisticsFrame(const crsfLinkStatistics_t* statsPtr, timeUs_t currentTimeUs)
{
    const crsfLinkStatistics_t stats = *statsPtr;
    lastLinkStatisticsFrameUs = currentTimeUs;
    switch (debugMode) {
    case DEBUG_CRSF_LINK_STATISTICS_UPLINK:
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 0, stats.Uplink_RSSI_1);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 1, stats.Uplink_RSSI_2);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 2, stats.Uplink_Link_quality);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_UPLINK, 3, stats.RF_Mode);
        break;
    case DEBUG_CRSF_LINK_STATISTICS_PWR:
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 0, stats.active_antenna);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 1, stats.Uplink_SNR);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_PWR, 2, stats.Uplink_TX_Power);
        break;
    case DEBUG_CRSF_LINK_STATISTICS_DOWN:
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 0, stats.Downlink_RSSI);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 1, stats.Downlink_Link_quality);
        DEBUG_SET(DEBUG_CRSF_LINK_STATISTICS_DOWN, 2, stats.Downlink_SNR);
        break;
    }
    const uint8_t rssidBm = stats.active_antenna ? stats.Uplink_RSSI_2 : stats.Uplink_RSSI_1;
    const uint8_t rssidBmScaled = scaleRange(crsfConvertRssidBm(rssidbm), 1, 100, 0, RSSI_MAX_VALUE);
    const uint8_t linkQuality = stats.Uplink_Link_quality + ((stats.RF_Mode == 2) ? 100 : 0);
    setRssiDirect(rssidBmScaled, RSSI_SOURCE_RX_PROTOCOL_CRSF);
    setLinkQualityDirect(linkQuality);
}
```

Reply:
i cleaned it up a bit 
your help is greatly appreciated. 
 
need to retest it all . and verify it sitll works